*Copyright (c) Cornac Authors. All rights reserved.*

*Licensed under the Apache 2.0 License.*

# Model Ensembling

<table class="tfo-notebook-buttons" align="left">
  <td>
    <a target="_blank" href="https://colab.research.google.com/github/PreferredAI/cornac/blob/master/tutorials/model_ensembling.ipynb"><img src="https://www.tensorflow.org/images/colab_logo_32px.png" />Run in Google Colab</a>
  </td>
  <td>
    <a target="_blank" href="https://github.com/PreferredAI/cornac/blob/master/tutorials/model_ensembling.ipynb"><img src="https://www.tensorflow.org/images/GitHub-Mark-32px.png" />View source on GitHub</a>
  </td>
</table>

This notebook provides an example of how to ensemble multiple recommendation models in Cornac.

Ensemble models is a technique that combines the predictions of multiple models to produce a single prediction. The idea is that by combining the predictions of multiple models, we can improve the overall performance of the recommendation system.

We will use the MovieLens 100K dataset and ensemble 2 models.

** Note: ** This notebook requires the `scikit-learn` package. 

## 1. Setup

### 1.1 Install required dependencies

In [1]:
! pip install seaborn scikit-learn

In [147]:
from IPython.display import display
import numpy as np
import pandas as pd

from cornac.datasets import movielens
from cornac.models import BPR, WMF
from cornac.eval_methods import RatioSplit
from cornac.metrics import Precision, Recall
from cornac.utils import cache
from cornac import Experiment

from sklearn import linear_model
from sklearn.ensemble import RandomForestRegressor

## 2. Prepare Experiment

### 2.1 Loading Dataset

First, we load the MovieLens 100K dataset.

In [145]:
data = movielens.load_feedback(variant="100K") # Load MovieLens Dataset

rs = RatioSplit(data, test_size=0.2, seed=42, verbose=True) # Split to train-test set to 80-20
train_set, test_set = rs.train_set, rs.test_set

rating_threshold = 1.0
exclude_unknowns = True
---
Training data:
Number of users = 943
Number of items = 1651
Number of ratings = 80000
Max rating = 5.0
Min rating = 1.0
Global mean = 3.5
---
Test data:
Number of users = 943
Number of items = 1651
Number of ratings = 19964
Number of unknown users = 0
Number of unknown items = 0
---
Total users = 943
Total items = 1651


### 2.2 Training BPR and WMF models

We will train two models: 

1. BPR (Bayesian Personalized Ranking)
2. WMF (Weighted Matrix Factorization)

In [146]:
bpr_model = BPR(k=50, max_iter=100, learning_rate=0.01, lambda_reg=0.001) # Initialize BPR model
wmf_model = WMF(k=50, max_iter=300, a=1.0, b=0.1, learning_rate=0.001, lambda_u=0.01, lambda_v=0.01,) # Initialize WMF model

models = [bpr_model, wmf_model]
metrics = [Precision(k=50), Recall(k=50)] # Set metrics for experiment

experiment = Experiment(rs, models, metrics, user_based=True).run() # Run Experiment to compare BPR model to WMF model individually


[BPR] Training started!

[BPR] Evaluation started!


Ranking: 100%|██████████| 940/940 [00:00<00:00, 1195.12it/s]



[WMF] Training started!


100%|██████████| 300/300 [00:29<00:00, 10.17it/s, loss=102] 


Learning completed!

[WMF] Evaluation started!


Ranking: 100%|██████████| 940/940 [00:00<00:00, 1389.07it/s]


TEST:
...
    | Precision@50 | Recall@50 | Train (s) | Test (s)
--- + ------------ + --------- + --------- + --------
BPR |       0.1846 |    0.5081 |    1.7945 |   0.7931
WMF |       0.1650 |    0.4909 |   29.7675 |   0.6808



Comparing Precision and Recall, both BPR and WMF are providing comparable results.

Let's move on to try to interpret these results by using the genres of movies that were recommended to us.

Generally, we could assume that if an individual likes a particular film genre like 'Romance', the recommender system should provide more of such 'Romance' films.

### 2.3 Interpreting Results

##### Creating a Movie Genre Dataframe

In [149]:
# Creating a dataframe of movies with its corresponding genres

# Download some information of MovieLens 100K dataset
item_df = pd.read_csv(
  cache("http://files.grouplens.org/datasets/movielens/ml-100k/u.item"), 
  sep="|", encoding="ISO-8859-1",
  names=["ItemID", "Title", "Release Date", "Video Release Date", "IMDb URL", 
         "unknown", "Action", "Adventure", "Animation", "Children's", "Comedy", 
         "Crime", "Documentary", "Drama", "Fantasy", "Film-Noir", "Horror", 
         "Musical", "Mystery", "Romance", "Sci-Fi", "Thriller", "War", "Western"]
).set_index("ItemID").drop(columns=["Video Release Date", "IMDb URL", "unknown"])

item_idx2id = list(train_set.item_ids) # create a item index to film ID mapping

# Let's take a look at an example of this dataframe
display(item_df.head(3))
     

Title Release Date  Action  Adventure  Animation  \
ItemID                                                                 
1        Toy Story (1995)  01-Jan-1995       0          0          1   
2        GoldenEye (1995)  01-Jan-1995       1          1          0   
3       Four Rooms (1995)  01-Jan-1995       0          0          0   

        Children's  Comedy  Crime  Documentary  Drama  Fantasy  Film-Noir  \
ItemID                                                                      
1                1       1      0            0      0        0          0   
2                0       0      0            0      0        0          0   
3                0       0      0            0      0        0          0   

        Horror  Musical  Mystery  Romance  Sci-Fi  Thriller  War  Western  
ItemID                                                                     
1            0        0        0        0       0         0    0        0  
2            0        0        0        0       0         1    0        0  
3            0        0        0        0       0         1    0        0

The `item_df` dataframe consists of all movie items with its corresponding genre attributes.

Further down below, we are going to filter this table with the recommendations that we get from the recommender system models we created to get a better sense of its performance.

##### Creating Training Data Dataframe

To get a sense of what data has been inserted into our model for training, let's count the genres of the training data used to train the model.

But first, let's create a `training_data_df` dataframe with all training data.

The training data consists of 80000 triplets of **User Index**, **Item Index** and **Rating** rows as seen in the dataset summary in Section 2.1.

In [151]:
# Create a training data dataframe
training_data_df = pd.DataFrame(zip(*train_set.uir_tuple)) # adding all training data into dataframe
training_data_df.columns = ['user_idx', 'item_idx', 'rating'] # adding column names to the data
training_data_df['item_id'] = training_data_df.apply(lambda row: item_idx2id[int(row['item_idx'])], axis=1) # adding item ids to the data by converting item index to item ids

# Let's view a sample of the training data dataframe
display(training_data_df.head(3))

user_idx  item_idx  rating item_id
0         0         0     4.0     381
1         1         1     3.0     602
2         2         2     4.0     431

##### Filtering Training Data

Let's filter based on a particular user to learn more about the user.

We set ``UIDX`` to user index **3**, and ``TOPK`` to **50**, to get the top 50 recommendations in each model for comparison.

In [152]:
UIDX = 3
TOPK = 50

In [153]:
# Filter training data (rating = 5.0 and user index = UIDX)
filter_df = training_data_df[(training_data_df['rating'] == 5.0) & (training_data_df['user_idx'] == UIDX)]
filter_df = item_df.loc[[int(item_id) for item_id in filter_df["item_id"]]] # get genres of movie items

# Group by Movie Genre and Sum by genres
filter_df = filter_df.select_dtypes(np.number).sum() 
filter_df = filter_df.to_frame("Sum") # Let's call that column 'Sum'

# Add a new column '%' for the percentage of individual genre sum compared to total sum
filter_df["%"] = filter_df["Sum"] / filter_df["Sum"].sum() * 100
filter_df["%"] = filter_df["%"].round(1)

# Let's see the training data genres, sums and percentages
print("Movies rated 5.0 by user index 3 in training data")
display(filter_df.sort_values("Sum", ascending=False))

Movies rated 5.0 by user index 3 in training data


Sum     %
Drama         71  24.9
Comedy        39  13.7
Romance       32  11.2
Action        30  10.5
Thriller      29  10.2
Adventure     19   6.7
War           15   5.3
Crime         12   4.2
Sci-Fi         9   3.2
Mystery        8   2.8
Children's     5   1.8
Horror         5   1.8
Musical        4   1.4
Film-Noir      2   0.7
Animation      2   0.7
Western        2   0.7
Documentary    1   0.4
Fantasy        0   0.0

As shown above in the training data, the top genres for user index 3 with movies rated 5.0 include 'Drama', 'Comedy', 'Romance', 'Action' and 'Thriller'.

Let's now compare them to the recommendations of the BPR and WMF models respectively.

##### Interpreting Recommendations of BPR, WMF Models

In [155]:
# Get the Top 5 Genres in filtered training data for user index 3
top_genres = filter_df.sort_values("Sum", ascending=False).head(5).index.tolist()
print("\nTop 5 Genres in training data:", top_genres)

# Get top K recommendations for BPR
bpr_recommendations, bpr_scores = bpr_model.rank(UIDX) # rank recommendations by score
bpr_topk = [item_idx2id[iidx] for iidx in bpr_recommendations[:TOPK]] # convert item indexes into item ids
bpr_df = item_df.loc[[int(iid) for iid in bpr_topk]] # filter the movie genre dataframe by item ids

# Let's view the top recommendations for BPR by top genres
display("BPR: Top recommendations", bpr_df[["Title"] + top_genres].head(10))

# Now, let's do likewise for WMF
wmf_recommendations, wmf_scores = wmf_model.rank(UIDX) # rank recommendations by score
wmf_topk = [item_idx2id[iidx] for iidx in wmf_recommendations[:TOPK]] # convert item indexes into item ids
wmf_df = item_df.loc[[int(iid) for iid in wmf_topk]] # filter the movie genre dataframe by item ids

# View the top recommendations for WMF
display("WMF: Top recommendations", wmf_df[["Title"] + top_genres].head(10))


Top 5 Genres in training data: ['Drama', 'Comedy', 'Romance', 'Action', 'Thriller']


'BPR: Top recommendations'

Title  Drama  Comedy  Romance  Action  \
ItemID                                                                       
313                         Titanic (1997)      1       0        1       1   
237                   Jerry Maguire (1996)      1       0        1       0   
402                           Ghost (1990)      0       1        1       0   
69                     Forrest Gump (1994)      0       1        1       0   
781                     French Kiss (1995)      0       1        1       0   
70      Four Weddings and a Funeral (1994)      0       1        1       0   
181              Return of the Jedi (1983)      0       0        1       1   
255        My Best Friend's Wedding (1997)      0       1        1       0   
215                 Field of Dreams (1989)      1       0        0       0   
300                   Air Force One (1997)      0       0        0       1   

        Thriller  
ItemID            
313            0  
237            0  
402            1  
69             0  
781            0  
70             0  
181            0  
255            0  
215            0  
300            1

'WMF: Top recommendations'

Title  Drama  Comedy  Romance  Action  \
ItemID                                                                  
483                 Casablanca (1942)      1       0        1       0   
269            Full Monty, The (1997)      0       1        0       0   
301                   In & Out (1997)      0       1        0       0   
248        Grosse Pointe Blank (1997)      0       1        0       0   
485               My Fair Lady (1964)      0       0        1       0   
531                      Shine (1996)      1       0        1       0   
51         Legends of the Fall (1994)      1       0        1       0   
282            Time to Kill, A (1996)      1       0        0       0   
1119    Some Kind of Wonderful (1987)      1       0        1       0   
107              Moll Flanders (1996)      1       0        0       0   

        Thriller  
ItemID            
483            0  
269            0  
301            0  
248            0  
485            0  
531            0  
51             0  
282            0  
1119           0  
107            0

Now that we have seen the top recommendations of the BPR and WMF models, let's do a comparison by taking a look at the genre distribution.

##### Comparing Models by Genre Distribution

In [156]:
# Let's introduce `combined_df` for comparison.
# This dataframe will be used to compare models by summing up genres from recommendations of different models
combined_df = pd.DataFrame({
    "Train Data %": filter_df["%"],
    "BPR Sum": bpr_df.select_dtypes(np.number).sum(), # group by genres, then get sum of each genre
    "WMF Sum": wmf_df.select_dtypes(np.number).sum() # likewise for WMF
})

# Get percentages of movie genre sums
combined_df['BPR %'] = combined_df['BPR Sum'] / combined_df['BPR Sum'].sum() * 100 
combined_df["WMF %"] = combined_df["WMF Sum"] / combined_df["WMF Sum"].sum() * 100

combined_df = combined_df.round(1) # round all 
combined_df = combined_df.sort_values("Train Data %", ascending=False)

# Let's take a look at the genre distribution by percentages
display("Train Data to Recommended % Distribution", combined_df[['Train Data %', 'BPR %', 'WMF %']])

'Train Data to Recommended % Distribution'

Train Data %  BPR %  WMF %
Drama                24.9   17.5   23.4
Comedy               13.7   17.5   14.0
Romance              11.2   16.7   15.9
Action               10.5   13.3    8.4
Thriller             10.2    9.2    7.5
Adventure             6.7    8.3    5.6
War                   5.3    6.7    6.5
Crime                 4.2    0.8    4.7
Sci-Fi                3.2    5.0    4.7
Mystery               2.8    1.7    2.8
Children's            1.8    1.7    1.9
Horror                1.8    0.8    0.0
Musical               1.4    0.0    1.9
Film-Noir             0.7    0.0    0.9
Animation             0.7    0.8    0.0
Western               0.7    0.0    0.9
Documentary           0.4    0.0    0.0
Fantasy               0.0    0.0    0.9

Now that we have seen the distribution of individual models, we are curious about what kind of distribution we will get from ensembling these models.

Let's see what happens when we ensemble these two models. 

## 2. Simple model ensembling by Borda Count

We will ensemble the two models using the Borda Count method. The Borda Count method is a simple voting method that ranks the items based on the sum of their ranks from each model.

Assuming that we have a list of 5 items, the Borda Count method works as follows:

1. For each model, rank the items from 1 to 5 based on the predicted scores.
2. Sum the ranks of each item across all models.
3. Sort the items based on the sum of their ranks.
4. The top-ranked item is the final recommendation.
5. Repeat the process for the next user.

Given the below example for a random user 123:

| Rank | Model 1 | Model 2 | Model 3 | Allocated Points (N - rank) |
|------|---------|---------|---------|-----------------------------|
| 1    | A       | D       | E       | 5 - 1 = 4                   |
| 2    | B       | C       | A       | 5 - 2 = 3                   |
| 3    | C       | A       | B       | 5 - 3 = 2                   |
| 4    | D       | B       | D       | 5 - 4 = 1                   |
| 5    | E       | E       | C       | 5 - 5 = 0                   |

The final ranking is as follows after summing them all up

| Item | Total Points     |
|------|------------------|
| A    | 4 + 2 + 3 = 9    |
| B    | 3 + 1 + 2 = 6    |
| D    | 1 + 4 + 1 = 6    |
| C    | 2 + 3 + 0 = 5    |
| E    | 0 + 0 + 4 = 4    |

New ranking: A > B, D > C > E


Lets implement this method below by using pandas `DataFrame` for data manipulation.

In [158]:
# Let's create a new dataframe to calculate ranking and borda count
rank_df = pd.DataFrame({
    "ItemID": item_idx2id,
})

total_items = len(rank_df) # 1651 items

# Obtain inverse ranks of the items based on the BPR score
rank_df["BPR Score"] = bpr_scores
rank_df["BPR Rank"] = rank_df["BPR Score"].rank(ascending=False).astype(int) # Get Rank where 1 = Top recommendation
rank_df["BPR Inverse Rank"] = total_items - rank_df["BPR Rank"] # Get inverse rank by calculating ('Total Item count' - 'Rank')

# Do likewise for WMF
rank_df["WMF Score"] = wmf_scores
rank_df["WMF Rank"] = rank_df["WMF Score"].rank(ascending=False).astype(int) # Get Rank where 1 = Top recommendation
rank_df["WMF Inverse Rank"] = total_items - rank_df["WMF Rank"] # Get inverse rank by calculating ('Total Item count' - 'Rank')

# Get Borda Count by summing up inverse ranks of BPR and WMF
rank_df["Borda Count"] = rank_df["BPR Inverse Rank"] + rank_df["WMF Inverse Rank"]

# Round decimal places for readability purposes
rank_df = rank_df.round(3)

# Now let's take a look at the table with Borda Count 
display(rank_df)

ItemID  BPR Score  BPR Rank  BPR Inverse Rank  WMF Score  WMF Rank  \
0       381      2.490       163              1488      0.650       956   
1       602     -0.859      1023               628      2.137       516   
2       431      0.580       594              1057      3.612       253   
3       875      1.578       348              1303      2.417       465   
4       182      1.479       377              1274      4.238       140   
...     ...        ...       ...               ...        ...       ...   
1646   1635     -2.334      1510               141     -0.167      1437   
1647   1650     -2.193      1431               220     -0.224      1476   
1648   1647     -2.250      1468               183     -0.160      1435   
1649   1663     -2.602      1604                47     -0.042      1370   
1650   1584     -2.172      1423               228      0.013      1307   

      WMF Inverse Rank  Borda Count  
0                  695         2183  
1                 1135         1763  
2                 1398         2455  
3                 1186         2489  
4                 1511         2785  
...                ...          ...  
1646               214          355  
1647               175          395  
1648               216          399  
1649               281          328  
1650               344          572  

[1651 rows x 8 columns]

Now that we have Borda Count, let's rerank this list and to provide the ensembled model's recommendation.

In [161]:
# Introduce reranked dataframe for borda count
reranked_df = rank_df.sort_values("Borda Count", ascending=False)

# Let's take a look at the ensembled top 5 recommendations and their respective ranks
display("Re-ranked Top K Item recommendations", reranked_df.head(5))



'Re-ranked Top K Item recommendations'

ItemID  BPR Score  BPR Rank  BPR Inverse Rank  WMF Score  WMF Rank  \
305    181      4.218         7              1644      5.685        12   
200     70      4.258         6              1645      5.332        24   
246    255      4.189         8              1643      5.233        27   
279    402      4.587         3              1648      5.033        37   
422    282      3.750        33              1618      5.786         8   

     WMF Inverse Rank  Borda Count  
305              1639         3283  
200              1627         3272  
246              1624         3267  
279              1614         3262  
422              1643         3261

The top recommendation ItemID **181**, was ranked **7** on BPR and **12** on WMF.

By doing Borda Count, we are able to aggregate model recommendations.

Next, let's add the recommendations into the genre distribution dataframe to compare its results to the base models.

In [162]:
borda_count_topk = reranked_df["ItemID"].values[:TOPK] # Get top K (50) Item IDs

borda_df = item_df.loc[[int(i) for i in borda_count_topk]] # Filter genre data frame by the top item IDs

# Add Borda Count results into 'combined_df' dataframe for comparison
combined_df["Borda Count Sum"] = borda_df.select_dtypes(np.number).sum() # group by genre, and calculate sum of each genre
combined_df["Borda Count %"] = combined_df["Borda Count Sum"] / combined_df["Borda Count Sum"].sum() * 100 # Calculate percentage of sum to total
combined_df["Borda Count %"] = combined_df["Borda Count %"].round(1) # rounding for readability purposes

# Let's take a look at the genre distribution of train data, BPR, WMF and the newly added Borda Count
display("Borda Count Recommendations Distribution", combined_df[["Train Data %", "BPR %", "WMF %", "Borda Count %"]])

'Borda Count Recommendations Distribution'

Train Data %  BPR %  WMF %  Borda Count %
Drama                24.9   17.5   23.4           22.1
Comedy               13.7   17.5   14.0           10.7
Romance              11.2   16.7   15.9           15.6
Action               10.5   13.3    8.4           11.5
Thriller             10.2    9.2    7.5            7.4
Adventure             6.7    8.3    5.6            8.2
War                   5.3    6.7    6.5            7.4
Crime                 4.2    0.8    4.7            3.3
Sci-Fi                3.2    5.0    4.7            5.7
Mystery               2.8    1.7    2.8            2.5
Children's            1.8    1.7    1.9            1.6
Horror                1.8    0.8    0.0            0.8
Musical               1.4    0.0    1.9            0.0
Film-Noir             0.7    0.0    0.9            0.8
Animation             0.7    0.8    0.0            0.8
Western               0.7    0.0    0.9            0.8
Documentary           0.4    0.0    0.0            0.0
Fantasy               0.0    0.0    0.9            0.8

In the next section, we will further add more models to the ensemble.

## 3. Adding more models to the Borda Count ensemble

We can easily add more models to the ensemble by training them and adding them. One approach is to train a model with different initializations using different random seeds. By adding multiple similar models of different random seeds (`seed=123`), some models could perform better for a set of users, while other models could perform better for another set of users.

By ensembling these models, we could potentially achieve better performance when combined.

Let's try adding a few more similar models with different random seed initializations.

In [163]:
# WMF models with different seeds
wmf_model_123 = WMF(name="WMF_123", k=50, max_iter=300, a=1.0, b=0.1, learning_rate=0.001, lambda_u=0.01, lambda_v=0.01, seed=123)
wmf_model_456 = WMF(name="WMF_456", k=50, max_iter=300, a=1.0, b=0.1, learning_rate=0.001, lambda_u=0.01, lambda_v=0.01, seed=456)
wmf_model_789 = WMF(name="WMF_789", k=50, max_iter=300, a=1.0, b=0.1, learning_rate=0.001, lambda_u=0.01, lambda_v=0.01, seed=789)
wmf_model_888 = WMF(name="WMF_888", k=50, max_iter=300, a=1.0, b=0.1, learning_rate=0.001, lambda_u=0.01, lambda_v=0.01, seed=888)
wmf_model_999 = WMF(name="WMF_999", k=50, max_iter=300, a=1.0, b=0.1, learning_rate=0.001, lambda_u=0.01, lambda_v=0.01, seed=999)

models = [wmf_model_123, wmf_model_456, wmf_model_789, wmf_model_888, wmf_model_999]

# Let's run an experiment to take a look at how different these models are, with just different random seeds!
experiment = Experiment(rs, models, metrics, user_based=True).run()


[WMF_123] Training started!


100%|██████████| 300/300 [00:25<00:00, 11.84it/s, loss=102] 


Learning completed!

[WMF_123] Evaluation started!


Ranking: 100%|██████████| 940/940 [00:00<00:00, 1958.09it/s]



[WMF_456] Training started!


100%|██████████| 300/300 [00:24<00:00, 12.40it/s, loss=103] 


Learning completed!

[WMF_456] Evaluation started!


Ranking: 100%|██████████| 940/940 [00:00<00:00, 2163.67it/s]



[WMF_789] Training started!


100%|██████████| 300/300 [00:24<00:00, 12.18it/s, loss=102] 


Learning completed!

[WMF_789] Evaluation started!


Ranking: 100%|██████████| 940/940 [00:00<00:00, 1995.85it/s]



[WMF_888] Training started!


100%|██████████| 300/300 [00:23<00:00, 12.85it/s, loss=103] 


Learning completed!

[WMF_888] Evaluation started!


Ranking: 100%|██████████| 940/940 [00:00<00:00, 2135.03it/s]



[WMF_999] Training started!


100%|██████████| 300/300 [00:23<00:00, 12.94it/s, loss=102] 


Learning completed!

[WMF_999] Evaluation started!


Ranking: 100%|██████████| 940/940 [00:00<00:00, 2208.80it/s]


TEST:
...
        | Precision@50 | Recall@50 | Train (s) | Test (s)
------- + ------------ + --------- + --------- + --------
WMF_123 |       0.1666 |    0.4968 |   25.5771 |   0.4849
WMF_456 |       0.1651 |    0.4894 |   24.3713 |   0.4380
WMF_789 |       0.1668 |    0.4945 |   24.8069 |   0.4756
WMF_888 |       0.1668 |    0.4963 |   23.5024 |   0.4451
WMF_999 |       0.1656 |    0.4937 |   23.3367 |   0.4303



Based on the results, we can see that even within the same model, the results can vary. 

Let's try ensembling all these models together into 1 single model by Borda Count, and look at its recommendations.

In [164]:
# Let's create a different dataframe to calculate ranking and borda count
rank_2_df = pd.DataFrame({
    "ItemID": item_idx2id,
})

# Add a column named 'Enhanced Borda Count'
rank_2_df["Enhanced Borda Count"] = 0

# Calculate the inverse rank for each of the models and accumulate them into the 'Enhanced Borda Count' column
for model in models:
    name = model.name
    recommendations, scores = model.rank(UIDX)
    rank_2_df[name + "_rating"] = scores
    rank_2_df[name + "_rank"] = rank_2_df[name + "_rating"].rank(ascending=False).astype(int)
    rank_2_df[name + "_inverse_rank"] = total_items - rank_2_df[name + "_rank"]
    rank_2_df["Enhanced Borda Count"] = rank_2_df["Enhanced Borda Count"] + rank_2_df[name + "_inverse_rank"]

# Round results for readability
rank_2_df = rank_2_df.round(3)

# Let's sort and view the top recommendations!
print("Model score calculation:")
display(rank_2_df[["WMF_123_inverse_rank", "WMF_456_inverse_rank", "WMF_789_inverse_rank", "WMF_888_inverse_rank", "WMF_999_inverse_rank", "Enhanced Borda Count"]].sort_values("Enhanced Borda Count", ascending=False).head(10))

Model score calculation:


WMF_123_inverse_rank  WMF_456_inverse_rank  WMF_789_inverse_rank  \
37                   1649                  1641                  1649   
670                  1642                  1636                  1631   
197                  1620                  1635                  1630   
156                  1630                  1611                  1642   
527                  1644                  1623                  1633   
305                  1576                  1645                  1638   
246                  1648                  1643                  1626   
386                  1625                  1620                  1611   
122                  1619                  1603                  1615   
322                  1635                  1640                  1614   

     WMF_888_inverse_rank  WMF_999_inverse_rank  Enhanced Borda Count  
37                   1650                  1649                  8238  
670                  1638                  1642                  8189  
197                  1623                  1650                  8158  
156                  1619                  1613                  8115  
527                  1568                  1645                  8113  
305                  1649                  1604                  8112  
246                  1639                  1538                  8094  
386                  1626                  1611                  8093  
122                  1644                  1603                  8084  
322                  1606                  1586                  8081

In [165]:
# Now, let's add them to the combined dataframe for comparison with earlier models
enhanced_borda_count_topk = list(rank_2_df.sort_values("Enhanced Borda Count", ascending=False)["ItemID"].values[:TOPK])
enhanced_borda_df = item_df.loc[[int(i) for i in enhanced_borda_count_topk]]

combined_df["Enhanced Borda Count Sum"] = enhanced_borda_df.select_dtypes(np.number).sum()
combined_df["Enhanced Borda Count %"] = combined_df["Enhanced Borda Count Sum"] / combined_df["Enhanced Borda Count Sum"].sum() * 100
combined_df["Enhanced Borda Count %"] = combined_df["Enhanced Borda Count %"].round(1)

# Let's compare the recommendation distribution
display("Combined Recommendations Distribution", combined_df[["Train Data %", "BPR %", "WMF %", "Borda Count %", "Enhanced Borda Count %"]])

'Combined Recommendations Distribution'

Train Data %  BPR %  WMF %  Borda Count %  Enhanced Borda Count %
Drama                24.9   17.5   23.4           22.1                    28.3
Comedy               13.7   17.5   14.0           10.7                     7.5
Romance              11.2   16.7   15.9           15.6                    10.4
Action               10.5   13.3    8.4           11.5                    12.3
Thriller             10.2    9.2    7.5            7.4                     9.4
Adventure             6.7    8.3    5.6            8.2                    10.4
War                   5.3    6.7    6.5            7.4                     5.7
Crime                 4.2    0.8    4.7            3.3                     3.8
Sci-Fi                3.2    5.0    4.7            5.7                     4.7
Mystery               2.8    1.7    2.8            2.5                     2.8
Children's            1.8    1.7    1.9            1.6                     2.8
Horror                1.8    0.8    0.0            0.8                     0.0
Musical               1.4    0.0    1.9            0.0                     0.0
Film-Noir             0.7    0.0    0.9            0.8                     0.9
Animation             0.7    0.8    0.0            0.8                     0.9
Western               0.7    0.0    0.9            0.8                     0.0
Documentary           0.4    0.0    0.0            0.0                     0.0
Fantasy               0.0    0.0    0.9            0.8                     0.0

Now that we have touched on simple borda count, let's see how we could use other methods and frameworks to do advanced model ensembling.

## 4. Advanced Model Ensembling

We could continue by thinking of this as a meta-learning problem. We could treat recommendations of each base model as features and train a meta-learner to predict the final recommendation.

This could be any ML model such as a Linear Regression, Random Forest, Gradient Boosting, or even a Neural Network.


In this example, we will use a simple Linear Regression model to predict the final recommendation.

In [168]:
# Lets get all the scores for the models trained in Part 3.

# First, lets create training and test data dataframes
training_df = pd.DataFrame(zip(*train_set.uir_tuple)) # Add 'User Index', 'Item Index', 'Rating' triples as records in dataframe
training_df.columns = ['user_idx', 'item_idx', 'ground_rating'] # Set column names

# Similar to train dataframe
test_df = pd.DataFrame(zip(*test_set.uir_tuple))
test_df.columns = ['user_idx', 'item_idx', 'ground_rating']
test_df['item_id'] = test_df.apply(lambda row: item_idx2id[int(row['item_idx'])], axis=1) # Add 'Item ID' column into dataframe by converting 'Item Index' to 'Item ID'

# For each model, we add individual predicted ratings by individual models to training and test dataframes
for model in models:
    name = model.name
    training_df[name + "_rating"] = training_df.apply(lambda row: model.score(int(row['user_idx']), int(row['item_idx'])), axis=1)
    test_df[name + "_rating"] = test_df.apply(lambda row: model.score(int(row['user_idx']), int(row['item_idx'])), axis=1)


X_train = training_df[['WMF_123_rating', 'WMF_456_rating', 'WMF_789_rating', 'WMF_888_rating', 'WMF_999_rating']]
y_train = training_df['ground_rating']
X_test = test_df[['WMF_123_rating', 'WMF_456_rating', 'WMF_789_rating', 'WMF_888_rating', 'WMF_999_rating']]

display("X_train", X_train.head(3)) # predicting ratings as features
display("y_train", y_train.head(3)) # ground truth ratings
display("X_test", X_test.head(3)) # test set

'X_train'

WMF_123_rating  WMF_456_rating  WMF_789_rating  WMF_888_rating  \
0        3.840917        4.167051        4.315118        4.399220   
1        2.303503        2.667568        2.825883        2.917715   
2        3.115823        3.776977        3.649920        3.028431   

   WMF_999_rating  
0        3.974400  
1        3.102975  
2        3.865470

'y_train'

0    4.0
1    3.0
2    4.0
Name: ground_rating, dtype: float64

'X_test'

WMF_123_rating  WMF_456_rating  WMF_789_rating  WMF_888_rating  \
0        1.193508        4.461263        1.642062        3.039706   
1        3.526569        3.566906        4.210879        3.870379   
2        1.772423        3.509660        2.643438        2.832230   

   WMF_999_rating  
0        2.172840  
1        3.917837  
2        1.559444

In [169]:
# Let's now fit into a Linear Regression model
regr = linear_model.LinearRegression()
regr.fit(X_train, y_train)

y_pred = regr.predict(X_test)

test_df["Linear Regression Rating"] = y_pred

# filter by predicted rating
test_df = test_df.sort_values("Linear Regression Rating", ascending=False)

top_item_ids = test_df[test_df['user_idx'] == UIDX]['item_id'].values[:TOPK]

linear_regression_df = item_df.loc[[int(i) for i in top_item_ids]]
combined_df["Linear Regression Sum"] = linear_regression_df.select_dtypes(np.number).sum()
combined_df["Linear Regression %"] = combined_df["Linear Regression Sum"] / combined_df["Linear Regression Sum"].sum() * 100
combined_df["Linear Regression %"] = combined_df["Linear Regression %"].round(1)

display("Combined Recommendations Distribution", combined_df[["Train Data %", "BPR %", "WMF %", "Borda Count %", "Enhanced Borda Count %", "Linear Regression %"]])

'Combined Recommendations Distribution'

Train Data %  BPR %  WMF %  Borda Count %  \
Drama                24.9   17.5   23.4           22.1   
Comedy               13.7   17.5   14.0           10.7   
Romance              11.2   16.7   15.9           15.6   
Action               10.5   13.3    8.4           11.5   
Thriller             10.2    9.2    7.5            7.4   
Adventure             6.7    8.3    5.6            8.2   
War                   5.3    6.7    6.5            7.4   
Crime                 4.2    0.8    4.7            3.3   
Sci-Fi                3.2    5.0    4.7            5.7   
Mystery               2.8    1.7    2.8            2.5   
Children's            1.8    1.7    1.9            1.6   
Horror                1.8    0.8    0.0            0.8   
Musical               1.4    0.0    1.9            0.0   
Film-Noir             0.7    0.0    0.9            0.8   
Animation             0.7    0.8    0.0            0.8   
Western               0.7    0.0    0.9            0.8   
Documentary           0.4    0.0    0.0            0.0   
Fantasy               0.0    0.0    0.9            0.8   

             Enhanced Borda Count %  Linear Regression %  
Drama                          28.3                 18.2  
Comedy                          7.5                 15.5  
Romance                        10.4                 13.6  
Action                         12.3                 11.8  
Thriller                        9.4                 10.9  
Adventure                      10.4                  5.5  
War                             5.7                  2.7  
Crime                           3.8                  1.8  
Sci-Fi                          4.7                  2.7  
Mystery                         2.8                  0.9  
Children's                      2.8                  4.5  
Horror                          0.0                  1.8  
Musical                         0.0                  4.5  
Film-Noir                       0.9                  0.0  
Animation                       0.9                  2.7  
Western                         0.0                  1.8  
Documentary                     0.0                  0.0  
Fantasy                         0.0                  0.9

### Applying Random Forest model



In [170]:
randomforest_model = RandomForestRegressor(n_estimators=200, random_state=42) # sklearn Random Forest model
randomforest_model.fit(X_train, y_train)

y_pred = randomforest_model.predict(X_test)

test_df["Random Forest Rating"] = y_pred

# Sort the items based on the score
test_df = test_df.sort_values("Random Forest Rating", ascending=False)

top_item_ids = test_df[test_df['user_idx'] == UIDX]['item_id'].values[:TOPK]

random_forest_df = item_df.loc[[int(i) for i in top_item_ids]]
combined_df["Random Forest Sum"] = random_forest_df.select_dtypes(np.number).sum()
combined_df["Random Forest %"] = combined_df["Random Forest Sum"] / combined_df["Random Forest Sum"].sum() * 100
combined_df["Random Forest %"] = combined_df["Random Forest %"].round(1)

display("Combined Recommendations Distribution", combined_df[["Train Data %", "BPR %", "WMF %", "Borda Count %", "Enhanced Borda Count %", "Linear Regression %", "Random Forest %"]])

'Combined Recommendations Distribution'

Train Data %  BPR %  WMF %  Borda Count %  \
Drama                24.9   17.5   23.4           22.1   
Comedy               13.7   17.5   14.0           10.7   
Romance              11.2   16.7   15.9           15.6   
Action               10.5   13.3    8.4           11.5   
Thriller             10.2    9.2    7.5            7.4   
Adventure             6.7    8.3    5.6            8.2   
War                   5.3    6.7    6.5            7.4   
Crime                 4.2    0.8    4.7            3.3   
Sci-Fi                3.2    5.0    4.7            5.7   
Mystery               2.8    1.7    2.8            2.5   
Children's            1.8    1.7    1.9            1.6   
Horror                1.8    0.8    0.0            0.8   
Musical               1.4    0.0    1.9            0.0   
Film-Noir             0.7    0.0    0.9            0.8   
Animation             0.7    0.8    0.0            0.8   
Western               0.7    0.0    0.9            0.8   
Documentary           0.4    0.0    0.0            0.0   
Fantasy               0.0    0.0    0.9            0.8   

             Enhanced Borda Count %  Linear Regression %  Random Forest %  
Drama                          28.3                 18.2             21.6  
Comedy                          7.5                 15.5             14.7  
Romance                        10.4                 13.6             15.7  
Action                         12.3                 11.8             11.8  
Thriller                        9.4                 10.9             12.7  
Adventure                      10.4                  5.5              5.9  
War                             5.7                  2.7              2.9  
Crime                           3.8                  1.8              0.0  
Sci-Fi                          4.7                  2.7              2.9  
Mystery                         2.8                  0.9              2.0  
Children's                      2.8                  4.5              2.9  
Horror                          0.0                  1.8              1.0  
Musical                         0.0                  4.5              2.9  
Film-Noir                       0.9                  0.0              0.0  
Animation                       0.9                  2.7              2.0  
Western                         0.0                  1.8              0.0  
Documentary                     0.0                  0.0              0.0  
Fantasy                         0.0                  0.9              1.0